Mathlib/LinearAlgebra/QuadraticForm/QuadraticModuleCat.lean

```diff
@@ -0,0 +1,164 @@
+  id_comp g := Hom.ext _ _ <| Isometry.id_comp g.toIsometry
+  comp_id f := Hom.ext _ _ <| Isometry.comp_id f.toIsometry
+  assoc f g h := Hom.ext _ _ <| Isometry.comp_assoc h.toIsometry g.toIsometry f.toIsometry
+
+-- TODO: if `Quiver.Hom` and the instance above were `reducible`, this wouldn't be needed.
+@[ext]
+lemma hom_ext {M N : QuadraticModuleCat.{v} R} (f g : M ⟶ N) (h : f.toIsometry = g.toIsometry) :
+    f = g :=
+  Hom.ext _ _ h
+
+/-- Typecheck a `QuadraticForm.Isometry` as a morphism in `Module R`. -/
+abbrev ofHom {X : Type v} [AddCommGroup X] [Module R X]
+    {Q₁ : QuadraticForm R X} {Q₂ : QuadraticForm R X} (f : Q₁ →qᵢ Q₂) :
+    of Q₁ ⟶ of Q₂ :=
+  ⟨f⟩
+
+@[simp] theorem toIsometry_comp {M N U : QuadraticModuleCat.{v} R} (f : M ⟶ N) (g : N ⟶ U) :
+    (f ≫ g).toIsometry = g.toIsometry.comp f.toIsometry :=
+  rfl
+
+@[simp] theorem toIsometry_id {M : QuadraticModuleCat.{v} R}  :
+    Hom.toIsometry (𝟙 M) = Isometry.id _ :=
+  rfl
+
+instance concreteCategory : ConcreteCategory.{v} (QuadraticModuleCat.{v} R) where
+  forget :=
+    { obj := fun M => M
+      map := @fun M N f => f.toIsometry }
+  forget_faithful :=
+    { map_injective := @fun M N => FunLike.coe_injective.comp <| Hom.toIsometry_injective _ _ }
+
+instance hasForgetToModule : HasForget₂ (QuadraticModuleCat R) (ModuleCat R) where
+  forget₂ :=
+    { obj := fun M => ModuleCat.of R M
+      map := fun f => f.toIsometry.toLinearMap }
+
+@[simp]
+theorem forget₂_obj (X : QuadraticModuleCat R) :
+    (forget₂ (QuadraticModuleCat R) (ModuleCat R)).obj X = ModuleCat.of R X :=
+  rfl
+
+@[simp]
+theorem forget₂_map (X Y : QuadraticModuleCat R) (f : X ⟶ Y) :
+    (forget₂ (QuadraticModuleCat R) (ModuleCat R)).map f = f.toIsometry.toLinearMap :=
+  rfl
+
+variable {X Y Z : Type v}
+variable [AddCommGroup X] [Module R X] [AddCommGroup Y] [Module R Y] [AddCommGroup Z] [Module R Z]
+variable {Q₁ : QuadraticForm R X} {Q₂ : QuadraticForm R Y} {Q₃ : QuadraticForm R Z}
+
+/-- Build an isomorphism in the category `QuadraticModuleCat R` from a
+`QuadraticForm.IsometryEquiv`. -/
+@[simps]
+def ofIso (e : Q₁.IsometryEquiv Q₂) : QuadraticModuleCat.of Q₁ ≅ QuadraticModuleCat.of Q₂ where
+  hom := ⟨e.toIsometry⟩
+  inv := ⟨e.symm.toIsometry⟩
+  hom_inv_id := Hom.ext _ _ <| FunLike.ext _ _ e.left_inv
+  inv_hom_id := Hom.ext _ _ <| FunLike.ext _ _ e.right_inv
+
+@[simp] theorem ofIso_refl : ofIso (IsometryEquiv.refl Q₁) = .refl _ :=
+  rfl
+
+@[simp] theorem ofIso_symm (e : Q₁.IsometryEquiv Q₂) : ofIso e.symm = (ofIso e).symm :=
+  rfl
+
+@[simp] theorem ofIso_trans (e : Q₁.IsometryEquiv Q₂) (f : Q₂.IsometryEquiv Q₃) :
+    ofIso (e.trans f) = ofIso e ≪≫ ofIso f :=
+  rfl
+
+end QuadraticModuleCat
+
+namespace CategoryTheory.Iso
+
+open QuadraticForm
+
+variable {X Y Z : QuadraticModuleCat.{v} R}
+
+/-- Build a `QuadraticForm.IsometryEquiv` from an isomorphism in the category
+`QuadraticModuleCat R`. -/
+@[simps]
+def toIsometryEquiv (i : X ≅ Y) : X.form.IsometryEquiv Y.form where
```

Comment:
I would rather see this definition (and lemmas below) in the same namespace as `IsometryEquiv`, like: `IsometryEquiv.ofIso`. Apart from this, the PR looks good to me.

Reply:
This is to match `Iso.toLinearEquiv`.

Reply:
Usually, I would try not to put too many specific things in the general category theory namespaces, but as you point it out, we already have `Iso.toLinearEquiv`... so

bors merge